File: python-worker/src/al_iyaal_worker/subtitles.py
import re
from dataclasses import dataclass
from pathlib import Path

SRT_TIME_RE = re.compile(
    r"\s*(?P<start>\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(?P<end>\d{2}:\d{2}:\d{2},\d{3})\s*"
)


@dataclass(slots=True)
class SubtitleEntry:
    index: int
    start_time: float
    end_time: float
    text: str


def parse_srt_timestamp(value: str) -> float:
    hh_mm, ms = value.rsplit(",", maxsplit=1)
    hours, minutes, seconds = [int(part) for part in hh_mm.split(":")]
    millis = int(ms)
    if minutes >= 60 or seconds >= 60 or millis >= 1000:
        raise ValueError(f"Invalid SRT timestamp: {value}")
    return hours * 3600 + minutes * 60 + seconds + millis / 1000


def parse_srt(content: str) -> list[SubtitleEntry]:
    normalized = content.replace("\r\n", "\n").replace("\r", "\n")
    if not normalized.strip():
        return []
    normalized = normalized.strip("\n")

    blocks = re.split(r"\n{2,}", normalized)
    entries: list[SubtitleEntry] = []
    for block_number, block in enumerate(blocks, start=1):
        lines = block.split("\n")
        if len(lines) < 3:
            raise ValueError(
                f"Invalid SRT cue block {block_number}: expected index, timestamp, and text."
            )

        time_match = SRT_TIME_RE.fullmatch(lines[1])
        if time_match is None:
            raise ValueError(
                f"Invalid SRT timestamp in cue block {block_number}: {lines[1]!r}"
            )

        try:
            index = int(lines[0])
        except ValueError:
            raise ValueError(
                f"Invalid SRT cue index in block {block_number}: {lines[0]!r}"
            ) from None
        if index < 0:
            raise ValueError(f"Invalid SRT cue index in block {block_number}: {index}")

        try:
            start_time = parse_srt_timestamp(time_match.group("start"))
            end_time = parse_srt_timestamp(time_match.group("end"))
        except ValueError as error:
            raise ValueError(f"Invalid SRT timestamp in cue block {block_number}: {error}") from error
        if end_time <= start_time:
            raise ValueError(
                f"Invalid SRT range in cue block {block_number}: end must be after start."
            )

        text = "\n".join(lines[2:]).strip()
        if not text:
            raise ValueError(f"Empty SRT cue text in block {block_number}.")

        entries.append(
            SubtitleEntry(
                index=index,
                start_time=start_time,
                end_time=end_time,
                text=text,
            )
        )

    return sorted(entries, key=lambda entry: (entry.start_time, entry.index))


def sidecar_srt_path(video_path: Path) -> Path:
    return video_path.with_suffix(".srt")


def sidecar_analysis_path(video_path: Path) -> Path:
    return video_path.with_suffix(".analysis.json")

File: src/features/editor/subtitles.ts
import type { SubtitleEntry } from "@/features/media/types";

const SRT_TIME_PATTERN = /^(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})$/;

const parseSrtTime = (value: string): number | null => {
  const [hh, mm, rest] = value.split(":");
  if (!hh || !mm || !rest) {
    return null;
  }
  const [ss, ms] = rest.split(",");
  if (!ss || !ms) {
    return null;
  }
  const values = [hh, mm, ss, ms].map(Number);
  if (values.some((part) => !Number.isFinite(part))) {
    return null;
  }
  const [hours = 0, minutes = 0, seconds = 0, milliseconds = 0] = values;
  if (minutes >= 60 || seconds >= 60 || milliseconds >= 1000) {
    return null;
  }
  return hours * 3600 + minutes * 60 + seconds + milliseconds / 1000;
};

const invalidCueError = (cueNumber: number, reason: string) =>
  new Error(`Invalid subtitle cue ${cueNumber}: ${reason}.`);

const parseSrtBlock = (block: string, blockIndex: number): SubtitleEntry => {
  const lines = block.split("\n");
  if (lines.length < 3) {
    throw invalidCueError(blockIndex + 1, "expected an index, timestamp range, and text");
  }

  const index = Number(lines[0]?.trim());
  if (!Number.isInteger(index) || index < 1) {
    throw invalidCueError(blockIndex + 1, "the cue index must be a positive integer");
  }

  const timeMatch = lines[1]?.trim().match(SRT_TIME_PATTERN);
  if (!timeMatch) {
    throw invalidCueError(index, "the timestamp range is malformed");
  }

  const startTime = parseSrtTime(timeMatch[1] ?? "");
  const endTime = parseSrtTime(timeMatch[2] ?? "");
  if (startTime === null || endTime === null) {
    throw invalidCueError(index, "the timestamp contains an invalid value");
  }
  if (endTime <= startTime) {
    throw invalidCueError(index, "the end time must be after the start time");
  }

  const text = lines.slice(2).join("\n").trim();
  if (!text) {
    throw invalidCueError(index, "the subtitle text is empty");
  }

  return {
    endTime,
    index,
    startTime,
    text,
  };
};

export const parseSrt = (content: string): SubtitleEntry[] => {
  const normalized = content.replace(/\r\n/g, "\n").replace(/\r/g, "\n").trim();
  if (!normalized) {
    return [];
  }

  const blocks = normalized.split(/\n{2,}/).filter((block) => block.trim().length > 0);
  const subtitles = blocks.map(parseSrtBlock);

  return subtitles.sort(
    (left, right) => left.startTime - right.startTime || left.index - right.index,
  );
};

export const formatTime = (seconds: number, maxDuration = seconds) => {
  const hh = Math.floor(seconds / 3600);
  const mm = Math.floor((seconds % 3600) / 60);
  const ss = Math.floor(seconds % 60);
  if (maxDuration >= 3600) {
    return `${hh}:${String(mm).padStart(2, "0")}:${String(ss).padStart(2, "0")}`;
  }
  return `${mm}:${String(ss).padStart(2, "0")}`;
};

export const findSubtitleAtTime = (subtitles: SubtitleEntry[], currentTime: number) => {
  let lower = 0;
  let upper = subtitles.length - 1;
  let candidateIndex = -1;

  while (lower <= upper) {
    const middle = Math.floor((lower + upper) / 2);
    const subtitle = subtitles[middle];
    if (subtitle && subtitle.startTime <= currentTime) {
      candidateIndex = middle;
      lower = middle + 1;
    } else {
      upper = middle - 1;
    }
  }

  const candidate = subtitles[candidateIndex];
  return candidate && currentTime <= candidate.endTime ? candidate : undefined;
};
